Build failure: fex on non-4K pagesize systems

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

Attempt to build the fex package on a non-4K-pagesize kernel.

### Can Hydra reproduce this build failure?

No, Hydra cannot reproduce this build failure.

### Link to Hydra build job

_No response_

### Relevant log output

```console
[2630/6983] Linking CXX executable FEXCore_Tests/FEXCore_Tests_Allocator
FAILED: [code=1] FEXCore_Tests/FEXCore_Tests_Allocator FEXCore/unittests/APITests/FEXCore_Tests_Allocator-b12d07c_tests.cmake /build/source/build/FEXCore/unittests/API>
: && /nix/store/wkcizi40n6zh5yiq0kvhyzw3yxrvlg2z-clang-wrapper-21.1.2/bin/clang++ -O3 -DNDEBUG -fomit-frame-pointer -flto=thin -fuse-ld=lld -fPIE -pie -Xlinker --depen>
<jemalloc>: Unsupported system page size
<jemalloc>: Unsupported system page size
<jemalloc>: Unsupported system page size
terminate called without an active exception
CMake Error at /nix/store/vqarvv18nq3cb5ca4m74kmwk1cyk83gd-catch2-3.11.0/lib/cmake/Catch2/CatchAddTests.cmake:82 (message):
  Error listing tests from executable
  '/build/source/build/FEXCore_Tests/FEXCore_Tests_Allocator':

    Result: Subprocess aborted
    Output: 

Call Stack (most recent call first):
  /nix/store/vqarvv18nq3cb5ca4m74kmwk1cyk83gd-catch2-3.11.0/lib/cmake/Catch2/CatchAddTests.cmake:234 (catch_discover_tests_impl)
```

### Additional context

It is expected that fex requires a 4K pagesize to _run_, however I would expect it to _compile_ fine so that I can _build_ the derivation on my Asahi host system normally to then _run_ it inside muvm.

I see this as a similar issue as cross-compilation, where unit tests need to be skipped if it can not be guaranteed that they will be able to run on the build host.

Furthermore the result when building fex on a 16K pagesize system should not be different than if it was built on a 4K pagesize system.

### System metadata

 - system: `"aarch64-linux"`
 - host os: `Linux 6.17.9, NixOS, 26.05 (Yarara), 26.05.20251204.835c620`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.95.0-pre20251203-dev_3fcfedc
System type: aarch64-linux
Additional system types: 
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/yuka/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/yuka/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/yuka/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/yuka/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/mngfv2r8fpzpa7csr9xsz31cmvpa1qjs-lix-2.95.0-pre20251203-dev_3fcfedc/share`
 - nixpkgs: `/nix/store/1l10zj1vfgmgdk2y257wqlj03ydh52vm-source`

### Notify maintainers

@andre4ik3 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/fe/fex/package.nix
+++ b/pkgs/by-name/fe/fex/package.nix
@@ -192,7 +192,6 @@ llvmPackages.stdenv.mkDerivation (finalAttrs: {
   cmakeFlags = [
     (lib.cmakeFeature "USE_LINKER" "lld")
     (lib.cmakeFeature "OVERRIDE_VERSION" finalAttrs.version)
-    (lib.cmakeBool "BUILD_TESTING" finalAttrs.finalPackage.doCheck)
     (lib.cmakeBool "BUILD_THUNKS" true)
     (lib.cmakeBool "BUILD_FEXCONFIG" withQt)
     (lib.cmakeFeature "X86_32_TOOLCHAIN_FILE" "${toolchain32}")
@@ -202,8 +201,21 @@ llvmPackages.stdenv.mkDerivation (finalAttrs: {
 
   strictDeps = true;
 
-  # Unsupported on non-4K page size kernels (e.g. Apple Silicon)
+  # Running the tests isn't supported on non-4K pagesize systems, but the build
+  # itself doesn't require 4K pagesize. So, to avoid breaking the build, enable
+  # checkPhase by default (so that the check inputs are included) and then
+  # manually disable it if we're running on a non-4K pagesize system.
   doCheck = true;
+  preConfigure = ''
+    if [ "$(getconf PAGESIZE)" != "4096" ]; then
+      echo "Disabling checkPhase due to non-4K pagesize environment"
+      unset doCheck
+      cmakeFlagsArray+=("-DBUILD_TESTING:BOOL=FALSE")
+    else
+      echo "Keeping checkPhase as-is"
+      cmakeFlagsArray+=("${lib.cmakeBool "BUILD_TESTING" finalAttrs.doCheck}")
+    fi
+  '';
 
   nativeCheckInputs = [ nasm ];
   checkInputs = [ catch2_3 ];

--- a/pkgs/by-name/fe/fex/package.nix
+++ b/pkgs/by-name/fe/fex/package.nix
@@ -95,13 +95,13 @@ let
 in
 llvmPackages.stdenv.mkDerivation (finalAttrs: {
   pname = "fex";
-  version = "2511";
+  version = "2512";
 
   src = fetchFromGitHub {
     owner = "FEX-Emu";
     repo = "FEX";
     tag = "FEX-${finalAttrs.version}";
-    hash = "sha256-CulENHssPkCXI+oyVKwf3GN5bjxUok2+AHsoOQ+Mchc=";
+    hash = "sha256-G61FdzNctTp8jarTcnBXd+MQpMxnPqd33hblvi9UXNo=";
 
     leaveDotGit = true;
     postFetch = ''

